### PR TITLE
feat(audit): add type-suppression budget guards to the audit pipeline

### LINKED ISSUE #873: Add type-suppression budget guards to the audit pipeline
## Background

Roadmap #864, quality-gate track. Strict TypeScript only protects
quality if suppressions don't accumulate: unbudgeted
`@ts-expect-error` / `any` turn the migration into typed-looking
untyped code. The repository already has a ratchet precedent — the
`bundleBudgets` mechanism in `audit/sync-manifest.json` enforced by
`audit-docs --check` (#837) — and this gate mirrors that shape so it
runs in the bare-node lane with no typescript dependency.

Blocked by #865 (toolchain and the audit-docs hook point must
exist). Landing this early is intentional: budgets start at zero, so
every later wave must either convert cleanly or raise a budget with an
explicit PR callout.

## Proposed change

- Forbid `@ts-ignore` outright (count must be 0; `@ts-expect-error`
  is the only allowed escape because it self-expires when the error
  disappears).
- Require every `@ts-expect-error` to carry a same-line reason
  comment, and count occurrences across `src/` and `tests/` against a
  budget recorded next to the existing `bundleBudgets` (same ratchet
  rule: raising a limit requires an explicit callout in the PR
  description; lowering is always allowed).
- Budget explicit `any` (annotations and `as any` casts) the same
  way. Where Biome's `lint/suspicious/noExplicitAny` can enforce this
  natively in the installed lane, prefer it and keep the bare-node
  count as the install-free backstop.
- Wire the counts into `scripts/audit-docs.mjs --check` (pure
  `node:` text scan — no typescript needed) so both CI lanes and the
  pre-push contract enforce them.

## Acceptance criteria

- `node scripts/audit-docs.mjs --check` fails when a `@ts-ignore` is
  introduced, when a `@ts-expect-error` lacks a reason, or when either
  budget is exceeded (each demonstrated by a unit test), and passes on
  the current tree.
- Budgets are recorded declaratively with the ratchet rule documented
  alongside, mirroring the `bundleBudgets` shape.
- The check runs in the bare-node `lint.yml` lane without any package
  install.
- Initial budgets equal the measured current counts (expected 0 at
  landing time; record the measured value, not an aspiration).

## Candidate files

- `scripts/audit-docs.mjs` (via its `src/` source if already
  converted), `audit/sync-manifest.json`
- `tests/` coverage for the new checks, `biome` config if
  `noExplicitAny` is enabled

## Source

Part of the **TypeScript migration** roadmap (#864),
wizard-resolved with the maintainer on 2026-06-11.

---

_Autopilot suitability: 4 / 5 -- higher is more autopilot-suitable; below the configured floor is human-oriented._

### PR DESCRIPTION
**Pull request by kurone-kito**
## Summary

Adds the type-suppression budget guard from the #864 quality-gate track: a `bundleBudgets`-style ratchet that keeps strict TypeScript honest by blocking suppression accumulation in the bare-node audit lane.

- **`@ts-ignore` is forbidden outright** — `@ts-expect-error` is the only allowed escape (it self-expires when the error disappears), and every occurrence must carry a same-line reason.
- **`@ts-expect-error` and explicit `any`** (`: any` annotations, `as any` casts, `<any>` arguments) across `src/**/*.mts` and `tests/**/*.mjs` are counted against budgets recorded declaratively in `audit/sync-manifest.json` (`typeSuppressionBudgets`), with the ratchet rule in its `description`: raising a limit requires an explicit PR callout; lowering is always allowed.
- **Initial budgets are the measured current counts: 0 and 0.**
- The collector is a pure function in `consistency-helpers` (audit-docs wires it via `globFiles` + `readText`), so the check runs install-free in the bare-node `lint.yml` lane and every failure mode is unit-tested.
- The explicit-`any` scan is lexer-accurate: it strips line/block comments, string/template contents, and regex literals (escape pairs consumed; regex detected by the last-significant-token heuristic with bracket classes), so prose like "Fail-safe: any invalid token" in comments and quotes inside regexes do not corrupt the count. Verified against all 64 currently scanned files.

Closes #873

## Verification

- 13 new unit tests in `tests/type-suppression-budgets.test.mjs` demonstrate each acceptance criterion: introduced `@ts-ignore` fails, reason-less `@ts-expect-error` fails, either budget exceeded fails, clean current-tree-shaped input passes, plus lexer regressions (escaped-backslash strings, quoted regex literals, division-vs-regex).
- `node scripts/audit-docs.mjs --check` passes on the current tree; the C1 critique pass independently confirmed each probe (`const blindProbe: any = 1;` appended to previously lexer-blind files) now fails the audit end-to-end.
- `pnpm run typecheck`, `pnpm run build:check`, full `node --test tests/*.mjs` (892/892), dprint/markdownlint/cspell — all green.

## Notes for review

- Biome's `lint/suspicious/noExplicitAny` (on via the recommended set) only **warns** in the installed lane, so this audit budget is the blocking enforcement in both CI lanes; the docs section states this explicitly.
- Disclosed residual: `${...}` interpolation code inside template literals is stripped with the template, so an explicit `any` inside an interpolation is not counted (stated in the stripper's comment; no such occurrence exists today).
- Directive tokens and the `any` matcher are assembled from string fragments so the scanner's own source, its emitted artifact, and the test fixtures never trip the scan they implement.

## Summary by CodeRabbit

* **New Features**
  * Added TypeScript suppression budget enforcement to audits: configurable limits for `@ts-expect-error` and explicit `any`, applied across source and test files; `@ts-ignore` is forbidden and `@ts-expect-error` requires a same-line reason; budgets are ratcheted and validated fail-closed.

* **Documentation**
  * Added docs describing budgets, counting rules, prohibition/requirements, and limit-adjustment process.

* **Tests**
  * Added comprehensive tests covering detection, counting, edge cases, and validation behavior.

### CODE REVIEW

**Comment by kurone-kito** [comment]
_claude-fable-c340413a: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
**Rejected** — chatgpt-codex-connector[bot] did not review HEAD 780039d3f4493a0544ca31dfefc728410cd91122 (Codex usage limit reached); this is not a completed review.

**Comment by kurone-kito** [comment]
**Accepted** — coderabbitai[bot] completed a review of HEAD 780039d3f4493a0544ca31dfefc728410cd91122 (COMMENTED, no actionable findings raised in threads); the summary walkthrough is noted.

File: src/scripts/consistency-helpers.mts [hunk 1]
```diff
@@ -126,3 +126,301 @@ export function collectRootMarkdownAllowlistViolations(
...
+  let lastWord = '';
+  while (i < text.length) {
+    const ch = text[i];
+    const next = text[i + 1];
+    if (state === 'code') {
+      if (ch === '/' && next === '/') {
+        state = 'line-comment';
+        i += 2;
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/consistency-helpers.mts:315]: **Accepted** — fixed in 4e6b9f6c4b978468fa1e8b8ec09543e1482e8d27: the stripper now emits one space when entering a line or block comment, so a comment between tokens can no longer splice them together; a regression test locks in the block-comment-between-as-and-any case.

File: src/scripts/consistency-helpers.mts [hunk 2]
```diff
@@ -126,3 +126,301 @@ export function collectRootMarkdownAllowlistViolations(
...
+      if (ch === '/' && next === '*') {
+        state = 'block-comment';
+        i += 2;
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/consistency-helpers.mts:321]: **Accepted** — fixed in 4e6b9f6c4b978468fa1e8b8ec09543e1482e8d27: the stripper now emits one space when entering a line or block comment, so a comment between tokens can no longer splice them together; a regression test locks in the block-comment-between-as-and-any case.

File: tests/type-suppression-budgets.test.mjs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    violations[0],
+    /3 explicit any occurrence\(s\) exceed the recorded budget of 2/,
+  );
+});
+
+test('does not count any-like prose in comments or strings', () => {
+  const files = [
+    {
+      path: 'src/scripts/sample.mts',
+      text: [
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply tests/type-suppression-budgets.test.mjs:140]: **Accepted** — fixed in 4e6b9f6c4b978468fa1e8b8ec09543e1482e8d27: the stripper now emits one space when entering a line or block comment, so a comment between tokens can no longer splice them together; a regression test locks in the block-comment-between-as-and-any case.

**Comment by kurone-kito** [comment]
advisory-wait: claude-fable-c340413a 4e6b9f6c4b978468fa1e8b8ec09543e1482e8d27 2026-06-11T12:38:00Z

File: src/scripts/consistency-helpers.mts [hunk 4]
```diff
@@ -126,3 +126,306 @@ export function collectRootMarkdownAllowlistViolations(
...
+        const at = line.indexOf(TS_EXPECT_ERROR_TOKEN, searchFrom);
+        if (at === -1) {
+          break;
+        }
+        tsExpectErrorCount += 1;
+        const trailing = line
+          .slice(at + TS_EXPECT_ERROR_TOKEN.length)
+          .replace(/^\s*(?:--|—|:)?\s*/u, '')
+          .trim();
+        if (trailing.length < 3) {
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/consistency-helpers.mts]: **Accepted** — fixed in e2fe10b74d2539dcc29dbb521fd573fc85ef425f: the check now flags only a truly absent reason (including a bare separator); any non-empty trailing text such as an issue reference passes. Two regression tests added.

File: scripts/consistency-helpers.mjs [hunk 5]
```diff
@@ -93,3 +93,268 @@ export function collectRootMarkdownAllowlistViolations(repoFiles, config) {
...
+        const at = line.indexOf(TS_EXPECT_ERROR_TOKEN, searchFrom);
+        if (at === -1) {
+          break;
+        }
+        tsExpectErrorCount += 1;
+        const trailing = line
+          .slice(at + TS_EXPECT_ERROR_TOKEN.length)
+          .replace(/^\s*(?:--|—|:)?\s*/u, '')
+          .trim();
+        if (trailing.length < 3) {
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply scripts/consistency-helpers.mjs]: **Accepted** — fixed in e2fe10b74d2539dcc29dbb521fd573fc85ef425f: the check now flags only a truly absent reason (including a bare separator); any non-empty trailing text such as an issue reference passes. Two regression tests added.

**Comment by kurone-kito** [comment]
advisory-wait: claude-fable-c340413a e2fe10b74d2539dcc29dbb521fd573fc85ef425f 2026-06-11T12:51:12Z

File: src/scripts/audit-docs.mts [hunk 6]
```diff
@@ -495,6 +501,26 @@ function checkRootMarkdownAllowlist(
...
+function checkTypeSuppressionBudgets(
+  config: TypeSuppressionBudgetConfig | null,
+) {
+  if (!config) {
+    return;
+  }
+  const globs = Array.isArray(config.globs)
+    ? config.globs.map((glob) => String(glob))
+    : [];
+  const files = uniqueSorted(globs.flatMap(globFiles)).map((path) => ({
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/audit-docs.mts]: **Accepted** — fixed in f24066fcc570bf6ed60f00e85d2ff92c8089d24e: a present budget entry with missing, invalid, or empty globs is now an audit error (fail closed) instead of silently scanning zero files.

**Comment by kurone-kito** [comment]
advisory-wait: claude-fable-c340413a f24066fcc570bf6ed60f00e85d2ff92c8089d24e 2026-06-11T13:04:55Z

File: src/scripts/consistency-helpers.mts [hunk 7]
```diff
@@ -126,3 +126,309 @@ export function collectRootMarkdownAllowlistViolations(
...
+// test fixture would count against the budget.
+const TS_IGNORE_TOKEN = `@ts-${'ignore'}`;
+const TS_EXPECT_ERROR_TOKEN = `@ts-${'expect'}-error`;
+// The explicit-`any` matcher is likewise assembled from fragments so its
+// own pattern text never trips the scan when this file is scanned.
+const ANY_TOKEN = 'any';
+const EXPLICIT_ANY_PATTERN = new RegExp(
+  `(?::\\s*${ANY_TOKEN}\\b|\\bas\\s+${ANY_TOKEN}\\b|<${ANY_TOKEN}>)`,
+  'g',
+);
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/consistency-helpers.mts]: **Accepted** — fixed in 40c1b11050d45efd89ddb31601f859490b26c6aa: the matcher now counts every standalone any word in stripped code (excluding property access and larger identifiers), so wrapped type arguments like Set<any[]> or Map<string, any> are counted; regression test added.

**Comment by kurone-kito** [comment]
advisory-wait: claude-fable-c340413a 40c1b11050d45efd89ddb31601f859490b26c6aa 2026-06-11T13:15:45Z

File: src/scripts/audit-docs.mts [hunk 8]
```diff
@@ -495,6 +501,37 @@ function checkRootMarkdownAllowlist(
...
+  // instead of silently passing a CI quality gate.
+  const globs = Array.isArray(config.globs)
+    ? config.globs
+        .map((glob) => String(glob))
+        .filter((glob) => glob.trim().length > 0)
+    : [];
+  if (globs.length === 0) {
+    errors.push(
+      `${String(config.id ?? 'type-suppression-budgets')}: globs must be a non-empty array of glob strings`,
+    );
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/audit-docs.mts]: **Accepted** — fixed in 68d7e32f18d4fb6bbc080a67e28cc4bd8871c4d9: non-string and blank glob entries are now rejected (the audit fails on the misconfiguration) instead of being coerced into pseudo-globs.

**Comment by kurone-kito** [comment]
advisory-wait: claude-fable-c340413a 68d7e32f18d4fb6bbc080a67e28cc4bd8871c4d9 2026-06-11T13:28:47Z

File: src/scripts/consistency-helpers.mts [hunk 9]
```diff
@@ -126,3 +126,313 @@ export function collectRootMarkdownAllowlistViolations(
...
+// own pattern text never trips the scan when this file is scanned.
+// It counts every standalone `any` word in stripped code — annotations,
+// casts, and type arguments in any nesting (`Set<any[]>`,
+// `Map<string, any>`, unions) — excluding only property access
+// (`.any`) and larger identifiers. Deliberately conservative: an
+// unusual identifier literally named `any` would over-count, which
+// fails loud for a budget gate rather than letting a wrapped type
+// argument bypass it.
+const ANY_TOKEN = 'any';
+const EXPLICIT_ANY_PATTERN = new RegExp(`(?<![.$\\w])${ANY_TOKEN}\\b`, 'g');
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/consistency-helpers.mts:161]: **Rejected** — this is the documented fail-loud trade-off chosen in the previous review round (which asked for the opposite: broader matching so wrapped type arguments cannot bypass the budget). An identifier literally named any does not exist in this codebase and would over-count toward a visible audit failure, not a silent bypass; satisfying both precision and recall here requires a real TypeScript parser, which contradicts the issue acceptance criterion that this check stays a pure node: text scan in the bare-node lane. Biome noExplicitAny remains the precise (non-blocking) signal in the installed lane. The trade-off is stated in the matcher comment.

File: src/scripts/consistency-helpers.mts [hunk 10]
```diff
@@ -126,3 +126,313 @@ export function collectRootMarkdownAllowlistViolations(
...
+  let explicitAnyCount = 0;
+
+  for (const file of files) {
+    const lines = String(file.text ?? '').split(/\r?\n/);
+    for (let index = 0; index < lines.length; index += 1) {
+      const line = lines[index];
+      if (line.includes(TS_IGNORE_TOKEN)) {
+        violations.push(
+          `${id}: ${file.path}:${index + 1} uses the forbidden ${TS_IGNORE_TOKEN} directive; use ${TS_EXPECT_ERROR_TOKEN} with a same-line reason instead`,
+        );
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/consistency-helpers.mts:210]: **Rejected** — intentional and fail-loud: directives only take effect in comments, but a string containing the directive token is so rare that raw-line scanning errs toward a visible audit failure rather than a silent bypass; the documented in-repo pattern for legitimate occurrences (this scanner and its tests) is fragment assembly, which is already demonstrated. Scanning comments only would require the inverse of the stripper and parser-level complexity that the bare-node acceptance criterion deliberately avoids.

**Comment by kurone-kito** [comment]
_claude-fable-c340413a: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-fable-c340413a: critique baseline — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
**Accepted** — coderabbitai[bot] completed its review cycle for HEAD 68d7e32f18d4fb6bbc080a67e28cc4bd8871c4d9: the single Major finding (non-string glob coercion) was fixed in 68d7e32 and confirmed by the bot in-thread; the updated summary walkthrough is noted.

File: src/scripts/audit-docs.mts (hunk 8 above)
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/audit-docs.mts]: **Accepted** — confirmation noted; the fix stands in 68d7e32f18d4fb6bbc080a67e28cc4bd8871c4d9.

**Comment by kurone-kito** [comment]
_claude-fable-c340413a: review triage snapshot — IDD automation marker. Do not edit._

File: src/scripts/audit-docs.mts (hunk 8 above)
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/audit-docs.mts]: **Accepted** — closing confirmation: the non-string glob rejection landed in 68d7e32f18d4fb6bbc080a67e28cc4bd8871c4d9 and coderabbitai marked it addressed; no further action.

**Comment by kurone-kito** [comment]
_claude-fable-c340413a: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
| Field | Value |
| --- | --- |
| Phase | F3 merged |
| Claim | claude-fable-c340413a / 46ee0832e031ddf9 |
| Branch | issue/873-add-type-suppression-budget-guards-audit |
| Last checked | 2026-06-11T13:49:13.183Z |
| Open blockers | none |
| Next action | F4 cleanup then F5 discover |
| Authoritative by | merge commit a2fc653 (matched head 68d7e32) |

**Comment by kurone-kito** [comment]
_claude-fable-c340413a: post-merge comment cleanup — IDD automation marker._

- status: applied (post-apply dry-run: clean)
- applied: 28 (operational markers, dispositioned bot summaries, resolved review thread comments, bot review parents)
- failed: 0
- skipped: 4 (policy: notices lacking completed-review signal / review parents without threads)
- viewer-cannot-minimize: 0
